Preprocessor.idr

```diff
@@ -0,0 +1,123 @@
+module Preprocessor
+
+import System.File.ReadWrite
+import System.File.Handle
+import System
+import Control.Monad.Error.Either
+import Control.Monad.Trans
+import Data.List1
+import Data.List
+import Data.String
+import Data.String.Extra
+
+record MimeRecord where
+  constructor MkMimeRecord
+  mime_type : String
+  mime_exts : List String
+
+record_to_declaration_name : MimeRecord -> String
+record_to_declaration_name mime = concat $ map go $ unpack mime.mime_type
+  where
+    go : Char -> String
+    go '+' = "_PLUS_"
+    go '.' = "_DOT_"
+    go chr = if isAlphaNum chr then cast (toUpper chr) else "_"
+
+tail : String -> String
+tail "" = ""
+tail xs = assert_total (strTail xs)
+
+cap_first : String -> String
+cap_first str =
+  case fromList (unpack str) of
+    Just (h ::: t) => pack $ forget (toUpper h ::: t)
+    Nothing => ""
+
+record_to_sub_type : MimeRecord -> String
+record_to_sub_type mime = tail $ snd $ break ('/' ==) mime.mime_type
+
+record_to_main_type : MimeRecord -> String
+record_to_main_type mime = concat $ map cap_first $ split (not . isAlphaNum) $ fst $ break ('/' ==) mime.mime_type
+
+parse_line : String -> Maybe MimeRecord
+parse_line line = do
+  let line = remove_comment line
+  (type ::: exts) <- List1.fromList $ filter (/= "") $ forget $ split isSpace line
+  pure (MkMimeRecord type exts)
+  where
+    remove_comment : String -> String
+    remove_comment str = trim (if isPrefixOf "#" str then tail str else str)
+
+read_line_or_eof : HasIO io => File -> EitherT FileError io (Maybe String)
+read_line_or_eof file = do
+  False <- lift (fEOF file)
+  | True => pure Nothing
+  map Just $ MkEitherT $ fGetLine file
+
+read_from_file : HasIO io => String -> io (Either FileError (List MimeRecord))
+read_from_file file = runEitherT $ do
+  handle <- MkEitherT {m=io} (openFile file Read)
```

Comment:
I prefer to close files in every program I write, which is important for long running programs and applications.
While this preprocessor is ending fast and OS is closing open file handles, I still would like to ask you to use `withFile` instead.

Reply:
done!